group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -2402,143 +2610,2090 @@ public void testOnNewMetadataImage() {
+            .setMemberId(newMemberId)
+            .setGenerationId(1)
+            .setProtocolType("consumer")
+            .setProtocolName("range")
+            .setSkipAssignment(supportSkippingAssignment)
+            .setErrorCode(Errors.NONE.code());
+
+        if (supportSkippingAssignment) {
+            expectedResponse
+                .setMembers(Collections.singletonList(
+                    new JoinGroupResponseData.JoinGroupResponseMember()
+                        .setMemberId(newMemberId)
+                        .setGroupInstanceId("group-instance-id")
+                        .setMetadata(protocols.find("range").metadata())
+                    ))
+                .setLeader(newMemberId);
+        }
+
+        GenericGroupMember updatedMember = group.member(group.staticMemberId("group-instance-id"));
+
+        assertEquals(expectedResponse, responseFuture.get(5, TimeUnit.SECONDS));
+
+        assertEquals(newMemberId, updatedMember.memberId());
+        assertEquals(Optional.of("group-instance-id"), updatedMember.groupInstanceId());
+        assertEquals(7000, updatedMember.rebalanceTimeoutMs());
+        assertEquals(4500, updatedMember.sessionTimeoutMs());
+        assertEquals(protocols, updatedMember.supportedProtocols());
+
+        assertEquals(1, group.size());
+        assertEquals(1, group.generationId());
+        assertTrue(group.isInState(STABLE));
+    }
+
+    @Test
+    public void testReplaceStaticMemberInStableStateWithUpdatedProtocolTriggersRebalance() throws Exception {
+
+        GroupMetadataManagerTestContext context = new GroupMetadataManagerTestContext.Builder()
+            .build();
+
+        JoinGroupRequestProtocolCollection protocols = new JoinGroupRequestProtocolCollection(0);
+
+        protocols.add(new JoinGroupRequestProtocol()
+            .setName("range")
+            .setMetadata(ConsumerProtocol.serializeSubscription(new ConsumerPartitionAssignor.Subscription(
+                Collections.singletonList("foo"))).array())
+        );
+
+        protocols.add(new JoinGroupRequestProtocol()
+            .setName("roundrobin")
+            .setMetadata(ConsumerProtocol.serializeSubscription(new ConsumerPartitionAssignor.Subscription(
+                Collections.singletonList("bar"))).array())
+        );
+
+        JoinGroupRequestData request = new JoinGroupRequestBuilder()
+            .withGroupId("group-id")
+            .withGroupInstanceId("group-instance-id")
+            .withMemberId(UNKNOWN_MEMBER_ID)
+            .withProtocolType("consumer")
+            .withProtocols(protocols)
+            .build();
+
+        JoinGroupResponseData response = context.joinGenericGroup(request, true, true);
+        assertEquals(Errors.NONE.code(), response.errorCode());
+
+        GenericGroup group = genericGroup(context, "group-id");
+
+        assertEquals(1, group.size());
+        assertEquals(1, group.generationId());
+        assertTrue(group.isInState(COMPLETING_REBALANCE));
```

Comment:
How do end up in this state here? Is there some code to advance the timer to complete the prepare phase?

Reply:
`sendGenericGroup...` methods send the request and return the future.
`joinGenericGroup...` methods invoke `sendGenericGroup` methods then advance the timer to move the group to completing rebalance state.

Reply:
With the latest changes, i renamed `joinGenericGroup...` to ``joinGenericGroupAsDynamicMemberAndCompleteJoin` and `joinGenericGroupAndCompleteJoin` to make it more explicit.

let me know if this is more readable.